Correct customer lifetime value calculation in customers

even-wei opened the pull request:
**Issue:**
The current calculation of customer lifetime value (CLV) in `customers` model is incomplete, leading to potentially inaccurate business insights.

**Solution:**
Modify the CLV calculation within `customers` to consider only 'completed' orders.

clkao commented:
Should we also check customer_order_pattern?

DaveFlynn commented:
## Description & motivation

The customer lifetime value (CLV) in customers model incorrectly included orders that were not yet completed, leading to potentially inaccurate business insights.
I modified the CLV calculation within customers to consider only "completed" orders.

## Type of Change

- [ ] New model
- [x ] Bugfix
- [ ] Refactoring
- [ ] Breaking change
- [ ] Documentation

## To-do before merge
- [ ]  Ask Sarah to review the customer segment impacts on marketing (see impact section below)

## Lineage DAG:





## Validation of models:

<details><summary>Value diff of customers</summary>

Modifying the calculation in the customers model to only include completed orders has changed 27% of the customer_lifetime_value column.


</details>



## Impact Considerations:





## Checklist:
<!---
This checklist is mostly useful as a reminder of small things that can easily be
forgotten – it is meant as a helpful tool rather than hoops to jump through.
Put an `x` in all the items that apply, make notes next to any that haven't been
addressed, and remove any items that are not relevant to this PR.
-->
- [ ] My pull request represents one logical piece of work.
- [ ] My commits are related to the pull request and look clean.
- [ ] My SQL follows the [dbt Labs style guide](https://github.com/dbt-labs/corp/blob/master/dbt_style_guide.md).
- [ ] `dbt build` completes successfuly and dbt tests pass (if not, detail why)
- [ ] I have materialized my models appropriately.
- [ ] I have added appropriate tests and documentation to any new models.
- [ ] I have updated the README file.
{%- if project.warehouse == 'redshift' %}
- [ ] I have added sort and dist keys to models materialized as tables.
- [ ] I have validated the SQL in any late-binding views.
{% endif %}

even-wei commented:
There is no change in `customer_order_pattern`.
I will add it and update the checklist.
![image](https://github.com/DataRecce/jaffle_shop_duckdb/assets/7878777/66d72e9e-89e1-4079-83a1-27748d9a3284)

even-wei commented:
Since the sources have been updated in the production environment, I've rebased the PR, rerun the checks, and updated the body. Most of the checks yield the same results, but I noticed that 99% of the value segments are classified as 'high value' (refer to the Top-k diff above). I plan to review the threshold next week.

popcornylu commented:
## Summary

PR #1 fixes the customer lifetime value (CLV) calculation in the `customers` model to only consider "completed" orders, addressing a critical business logic issue. The change modifies **1 SQL file** and impacts **2 downstream models** (`customer_segments` and `customer_order_pattern`).

---

## Key Changes

- **Modified Model**: `customers` table - Updated CLV calculation logic to filter for "completed" orders only
- **Data Impact**: `customer_lifetime_value` column shows **-32% reduction** in mean value (**$2,758.60 → $1,871.77**)
- **Quality Issue**: The column now contains **1 null value** (was previously 100% not-null)
- **Distinctness**: Lost **96 unique values** in `customer_lifetime_value` (**-6.9%** distinct proportion)
- **Verification**: Row counts remain stable at **1,856 records** across all affected models

---

## Impact Analysis

```mermaid
graph LR
    stg_customers["stg_customers<br/>(view)"]:::unchanged
    stg_payments["stg_payments<br/>(view)"]:::unchanged
    stg_orders["stg_orders<br/>(view)"]:::unchanged
    customers["customers<br/>(table)"]:::modified
    customer_segments["customer_segments<br/>(table)"]:::impacted
    customer_order_pattern["customer_order_pattern<br/>(table)"]:::impacted

    stg_customers --> customers
    stg_orders --> customers
    stg_payments --> customers
    customers --> customer_segments
    customers --> customer_order_pattern

    classDef modified fill:#fff3cd,stroke:#ffc107,color:#000000
    classDef impacted fill:#ffffff,stroke:#ffc107,color:#000000
    classDef unchanged fill:#ffffff,stroke:#d3d3d3,color:#999999
```

- ⚠️ **customers**: Logic change in CLV calculation - now excludes non-completed orders, introducing **1 null value** in the lifetime value column
- ⚠️ **customer_segments**: Downstream table impacted by CLV changes; remains stable with **1,856 rows** but may reflect lower segmentation values
- ⚠️ **customer_order_pattern**: Indirectly impacted; row counts stable at **1,856** but patterns may shift with corrected CLV data
- ✅ **Data integrity**: No schema changes; all 6 upstream/downstream models structurally intact
- ✅ **Row counts**: All affected tables maintain consistent row counts across environments

---

## ☑️ Checklist

No checks configured in the PR.

---

## 🔍 Suggested Actions

- [ ] **Investigate the null value**: Review why `customer_lifetime_value` now contains **1 null** record after the CLV calculation change and ensure it's expected behavior
- [ ] **Validate downstream impact**: Verify that `customer_segments` and `customer_order_pattern` produce correct results with the **-32% reduction** in CLV values
- [ ] **Add data quality checks**: Implement validation checks for not-null constraints on `customer_lifetime_value` and referential integrity with downstream models
- [ ] **Review business logic**: Confirm that filtering to "completed" orders only aligns with business requirements for customer segmentation and order patterns
- [ ] **Address PR comment**: Community member flagged `customer_order_pattern` - ensure this model produces accurate patterns with corrected CLV calculation
